### DISCUSSION TITLE
Accessing all `SavedCollection` posts

### DISCUSSION THREAD

**Discussion by diligiant** [sbertix/Swiftagram#245] [Q&A] [answered]
I'm having troubles using `Endpoint.saved.collection(id)` as the received value json doesn't fit the types. But it could be very well me ;) so I thought I'd ask before creating an issue.

json is `[saveMediaResponse: [items: [media: [_the_post_]`

`Endpoint.saved.collection(id)` returns a `SavedCollection.Unit` (publisher that is)

then
```
SavedCollection.Unit {
	.collection: SavedCollection? // [“savedMediaResponse”]
```
then
```
SavedCollection {
	.items: [Media]? // [“items”]
```
at which point I'm stuck as `Media` doesn't have the `media` getter that `Media.Unit` has.

I checked every combo but this has been fruitless ;)

Of course with Wrapped, there is always a workaround!
```
public extension Media {
  var media: Media? { self["media"].optional().flatMap(Media.init) }
}
```

**Comment by sbertix** [comment, marked as answer]
Hey @diligiant,

Unfortunately I'm out of town, and I'm only now having the chance to catch up with everything. 

If I got this correctly, I think the best solution would be extending `items` definition like this.

```swift
var items: [Media]? {
   self["saveMediaResponse"]["items"].array().compactMap {
      $0.media.optional().flatMap(Media.init)
			?? $0.optional().flatMap(Media.init)
   }
}
```

(I haven't checked the actual response model, btw, so apologies if it's not 100%: I'm just basing it off your post)

**Reply by diligiant** [reply]
thank you for your answer.

So you would like to put `items` directly into `SavedCollection.Unit` instead of following the "breadcrumb" (`.collection` `.items` `.media`)? Works for me ;)
Should I issue a PR?

**Reply by sbertix** [reply]
I usually consider 3 things when building `Specialized` models and their properties.

- are they universal? Do I really need a `Post.User`, a `Follower.User`, and a "`User.User`"? They represent the same thing just with a different level of granularity → **"one model to rule them all"**
- can the user expect different responses to behave the same? _Collections_ and _autocollections_ have slightly different models and "entirely" different endpoints, but they are undistinguishable from each other in the frontend client. Users expect them to behave consistently (another example would be "old posts" in the timeline and the way they're handled in **Swiftagram**) → **consistency** 
- am I overcomplicating things? What sense does it make having a `media` property on a `Media` instance? And what's the point of having a `MediaWrapper` having a `media: Media` property when all it does is unwrapping that `Media` instance? → **avoid redundancy** 

Under these assumptions, encompassing all `items` scenarios directly inside of `SavedCollection` makes the most sense. The property's already there: it should simply be tweaked to make sure it works under all circumstances. 

And yes, feel free to push a PR with this fix 😊 I'm sorry I haven't been able to get to it myself, I've been out of town for the past week.
